File: Top-Interview-Questions-Leetcode/Easy-Collections/linked-list/BT-level-order-traversal.py
from queue import Queue
from typing import List, Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if not root:
            return []
        
        ans = []
        queue = Queue()
        queue.put(root)
        
        while not queue.empty():
            sz = queue.qsize()
            vals = []
            
            for _ in range(sz):
                node = queue.get()
                
                vals.append(node.val)
                
                if node.left:
                    queue.put(node.left)
                if node.right:
                    queue.put(node.right)
            
            ans.append(vals)
        
        return ans

File: Top-Interview-Questions-Leetcode/Easy-Collections/linked-list/BT-level-order-traversal.cpp
#include <iostream>
#include <vector>
#include <queue>

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    std::vector<std::vector<int>> levelOrder(TreeNode* root) {
        if (!root) return {};
        
        std::vector<std::vector<int> > ans;
        std::queue<TreeNode*> q;
        q.push(root);
        
        while (!q.empty()) {
            int sz = q.size();
            std::vector<int> vals;
            
            while (sz--) {
                TreeNode* node = q.front();
                q.pop();
                
                vals.push_back(node->val);
                
                if (node->left) q.push(node->left);
                if (node->right) q.push(node->right);
            }
            
            ans.push_back(vals);
        }
        
        return ans;
    }
};